flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1195,11 +1196,16 @@ public void stopCheckpointScheduler() {
 				currentPeriodicTrigger = null;
 			}
 
-			for (PendingCheckpoint p : pendingCheckpoints.values()) {
+			// take a snapshot of pendingCheckpoints to clear them and prevent aborting them twice
+			// in case of subsequent call of stopCheckpointScheduler()
+			List<PendingCheckpoint> pendingCheckpointsSnapshot =
+				new ArrayList<>(pendingCheckpoints.values());
+			pendingCheckpoints.clear();
+
+			for (PendingCheckpoint p : pendingCheckpointsSnapshot) {
```

Comment:
Just as a note, I guess we could have solved the problem also by changing `JobMaster.java:965-975` into a `handleAsync` part because the problem seems to be that by aborting a `PendingCheckpoint` we fail the savepoint future which is then directly triggers the execution of the `exceptionally` handler. By decoupling the `exceptionally` handler from the `stopCheckpointScheduler` call, there should not be any concurrency problem.

Reply:
Agree, this fix will solve the original problem without other changes.